### ISSUE TITLE
Support for arrays of primitive types

### ISSUE THREAD

**Issue by StefanFabian** [dotnet/efcore#6487]
Is support for array properties planned?
If it is anything you can say about when to expect it?
Being able to save arrays of strings etc. seems pretty crucial to me.

**Comment by roji** [comment]
I'm not sure what this feature means in general for EFCore (or whether array properties are supported in SqlServer/Sqlite), but support for array properties has been added to the Npgsql provider for PostgreSQL (see https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/15). Note that it's in the dev branch (so planned for 1.1.0) but not yet released). No special support was needed from EFCore in order to support this.

**Comment by StefanFabian** [comment]
SQLite itself does not support arrays.
Since EFCore seems to be the preferred way to save application data on the disk, I wanted to move over from SQLite.Net-PCL which also seems to get less and less support/updates.
However my class contains string arrays. Not a problem for the pcl but EFCore doesn't like it.

I found that issue of yours but I wasn't sure if it had anything to do with my problem.
Can I use the  PostgreSQL version for UWP?
Does it ship with the anniversary update like sqlite or do I have to ship the dll myself?

**Comment by roji** [comment]
UWP isn't supported yet, although it's mainly due to lack of interest - I'm hoping adding support wouldn't be a very big deal (see https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/15).

Npgsql doesn't ship with any version of Windows and is available like any other nuget library.

**Comment by roji** [comment]
BTW, if SQLite doesn't support arrays, what kind of support exactly do you expect from EFCore here? I don't know SQLite.Net-PCL at all but how can it support array properties if SQLite doesn't?

**Comment by StefanFabian** [comment]
You can provide a Serializer that takes the object and returns a byte blob or takes a byte blob and returns an object.
I'm also using this to serialize an object property which - if I understand correctly - I wouldn't have to with EFCore and that'd be a big plus.

**Comment by roji** [comment]
You can do the equivalent thing in EFCore by exposing a byte[] property backed by your array of strings, and instructing EFCore to ignore the array property. EFCore only knows about the blob property, which handles conversions to and from the array property.

Regarding your object property, EFCore can serialize that for you into another table, creating a relationship between the two classes/tables. This is very different from serializing an object into a property. You may also want to take a look at #246, which tracks complex types (but is not implemented).

Again, for a PostgreSQL-specific angle, Npgsql allows you to map arbitrary objects to PostgreSQL "composite" types, which allows you to save an object as a single, simple property.

**Comment by StefanFabian** [comment]
Yeah I guess I could do that but it's kind of an ugly solution.

I like the approach of EFCore with object properties because I'm using an abstract base class as object property which makes serialization a bit complicated since it needs to save the type information and that can be quite annoying with .NET native.

**Comment by roji** [comment]
@StefanFabian you may want to look at [inheritance in EFCore](http://ef.readthedocs.io/en/latest/modeling/inheritance.html). Other than that I'm not sure what else EFCore can help you with...

**Comment by StefanFabian** [comment]
Yeah that's what I planned thanks for the link.
Guess I'll just create another subclass that is only visible to the database and handles the byte[] to string[] conversion and make the string[] properties ignored in the modelbuilder.

**Comment by StefanFabian** [comment]
Okay I solved the problem with the string array by using a class wrapper.

However it does not seem to like the abstract property.
So my Entry is roughly like the following:

`class Entry {
    public AbstractBase Property { get; set; }
}`

`class AbstractBase {  
    [Key]  
    public int Id { get; set; }  
}`

`class  Implementation : AbstractBase {
    public string AdditionalInfo { get; set; }
}`

The migration is created correctly and when adding my stuff to the database it works just fine. However when reading it from the database Entry.Property is always null.
I've checked the database manually and it has foreign keys for AbstractBase and the table for AbstractBase is filled.

Any idea what I did wrong?
